--- a/src/icedtea-sound/acinclude.m4
+++ b/src/icedtea-sound/acinclude.m4
@@ -527,7 +527,7 @@
     }
 }]
 EOF
-  if $JAVAC -cp . $JAVACFLAGS -source 5 -target 5 $CLASS >&AS_MESSAGE_LOG_FD 2>&1; then
+  if $JAVAC -cp . $JAVACFLAGS -source 6 -target 6 $CLASS >&AS_MESSAGE_LOG_FD 2>&1; then
     if $JAVA -classpath . $BYTECODE >&AS_MESSAGE_LOG_FD 2>&1; then
       it_cv_jdk_works=yes;
     else
--- a/src/icedtea-sound/configure
+++ b/src/icedtea-sound/configure
@@ -4601,7 +4601,7 @@
     }
 }
 EOF
-  if $JAVAC -cp . $JAVACFLAGS -source 5 -target 5 $CLASS >&5 2>&1; then
+  if $JAVAC -cp . $JAVACFLAGS -source 6 -target 6 $CLASS >&5 2>&1; then
     if $JAVA -classpath . $BYTECODE >&5 2>&1; then
       it_cv_jdk_works=yes;
     else
