--- a/src/java.smartcardio/unix/classes/sun/security/smartcardio/PlatformPCSC.java
+++ b/src/java.smartcardio/unix/classes/sun/security/smartcardio/PlatformPCSC.java
@@ -48,6 +48,7 @@ class PlatformPCSC {
 
     private final static String PROP_NAME = "sun.security.smartcardio.library";
 
+    private final static String LIB0 = "libpcsclite.so.1";
     private final static String LIB1 = "/usr/$LIBISA/libpcsclite.so";
     private final static String LIB2 = "/usr/local/$LIBISA/libpcsclite.so";
     private final static String PCSC_FRAMEWORK = "/System/Library/Frameworks/PCSC.framework/Versions/Current/PCSC";
@@ -100,6 +101,7 @@ class PlatformPCSC {
         if (lib.length() != 0) {
             return lib;
         }
+#if 0
         lib = expand(LIB1);
         if (new File(lib).isFile()) {
             // if LIB1 exists, use that
@@ -134,6 +136,10 @@ class PlatformPCSC {
             return lib;
         }
         throw new IOException("No PC/SC library found on this system");
+#endif
+	// let dlopen do the work
+	lib = LIB0;
+	return lib;
     }
 
     private static native void initialize(String libraryName);
