--- a/src/java.base/share/classes/java/lang/Float.java	2018-03-13 17:46:28.000000000 +0100
+++ b/src/java.base/share/classes/java/lang/Float.java	2018-05-01 15:35:55.626369726 +0200
@@ -92,7 +92,7 @@
      * hexadecimal floating-point literal {@code 0x0.000002P-126f}
      * and also equal to {@code Float.intBitsToFloat(0x1)}.
      */
-    public static final float MIN_VALUE = 0x0.000002P-126f; // 1.4e-45f
+    public static final float MIN_VALUE = Float.intBitsToFloat(0x1);
 
     /**
      * Maximum exponent a finite {@code float} variable may have.  It
