--- src/hotspot/os/linux/os_linux.cpp~	2024-12-31 07:26:39.000000000 +0100
+++ src/hotspot/os/linux/os_linux.cpp	2025-01-05 10:26:10.266071715 +0100
@@ -2985,6 +2985,10 @@
 // Note that the value for MAP_FIXED_NOREPLACE differs between architectures, but all architectures
 // supported by OpenJDK share the same flag value.
 #define MAP_FIXED_NOREPLACE_value 0x100000
+// on ALPHA, that triggers with 0x1000000 == 0x100000. why?
+#if defined(ALPHA) && defined(MAP_FIXED_NOREPLACE)
+#undef MAP_FIXED_NOREPLACE
+#endif
 #ifndef MAP_FIXED_NOREPLACE
   #define MAP_FIXED_NOREPLACE MAP_FIXED_NOREPLACE_value
 #else
